examples/imagenet.py

```diff
@@ -0,0 +1,363 @@
+import os
+import warnings
+
+import hydra
+import torch
+import torch.nn as nn
+import torch.backends.cudnn as cudnn
+import torch.distributed as dist
+import torchvision.models as models
+import torchvision.datasets as datasets
+import torchvision.transforms as transforms
+from torch.optim import Adam
+
+best_acc1 = 0
+
+
+@hydra.main(config_name="imagenetconf")
+def main(cfg):
+    if cfg.seed is not None:
+        random.seed(cfg.seed)
+        torch.manual_seed(cfg.seed)
+        cudnn.deterministic = True
+        warnings.warn(
+            "You have chosen to seed training. "
+            "This will turn on the CUDNN deterministic setting, "
+            "which can slow down your training considerably! "
+            "You may see unexpected behavior when restarting "
+            "from checkpoints."
+        )
+    if cfg.gpu is not None:
+        warnings.warn(
+            "You have chosen a specific GPU. This will completely "
+            "disable data parallelism."
+        )
+    if cfg.dist_url == "env://" and cfg.world_size == -1:
+        cfg.world_size = int(os.environ["WORLD_SIZE"])
+
+    global best_acc1
+
+    if cfg.gpu is not None:
+        print("Use GPU: {} for training".format(cfg.gpu))
+
+    if cfg.distributed:
+        if cfg.dist_url == "env://" and cfg.rank == -1:
+            cfg.rank = int(os.environ["RANK"])
+        if cfg.multiprocessing_distributed:
+            cfg.rank = cfg.rank * cfg.ngpus_per_node + gpu
+        dist.init_process_group(
+            backend=cfg.dist_backend,
+            init_method=cfg.dist_url,
+            world_size=cfg.world_size,
+            rank=cfg.rank,
+        )
+    if cfg.pretrained:
```

Comment:
Not that we can handle this now, but just putting this out there as a general statement:

Statements conditioned solely on the config often show up in main files and IMO reduce the readability of the code.

My wish for the future (once we have the hammers) would be to push this logic into the config. There are many pros:
1. `main.py` is shorter and cleaner with less nesting or extraneous code.
2. `main.py` can be more general. This implies greater extensibility for users, and less duplicate code.
3. As the config is resolved in real time through interpolation / logic resolution, you get to see the 'final' config all in one structure before it is run. This is very powerful in a sense that we now see the outcome of all the intermediate logic that happens in main before training begins.

Obviously this won't work for every case, but there are many (like this one) where I think it would be trivial.

Reply:
I completely agree with you, and I think it'd be really nice to get rid of as many conditional statements as possible. I think a combination of using configs and refactoring should do the job well.